Add documentation for running the panel in Docker (#113)
pelican-dev/docs: docs/panel/advanced/docker.mdx

> # Docker
> Pelican provides pre-built Docker images via GitHub Packages. `ghcr.io/pelican-dev/panel:latest` is the current latest release, and `ghcr.io/pelican-dev/panel:main` is built automatically from the current `main` branch. Deploying the panel in Docker is still a work in progress. While the plan is to make Docker the preferred installation method, we currently recommend the [standard deployment instructions](/docs/panel/getting-started)

Reviewer: ```suggestion
# Docker

Pelican provides pre-built Docker images via GitHub Packages. `ghcr.io/pelican-dev/panel:latest` is the current latest release, and `ghcr.io/pelican-dev/panel:main` is built automatically from the current `main` branch. Deploying the panel in Docker is still a work in progress. While the plan is to make Docker the preferred installation method, we currently recommend the [standard deployment instructions](/docs/panel/getting-started)
```

This guide requires Docker CE. (Docker Compose has been included in the Docker CLI since v2. Docker Compose v1 is unsupported.) For instructions on installing and configuring Docker, see the [installation guide](/docs/guides/docker).

> ## Basics
> The easiest deployment method is using the standard `compose.yml` file.

Reviewer: ```suggestion
## Basics

The easiest deployment method is using the standard `compose.yml` file.
```

This configuration includes an integrated web server that will automatically obtain SSL certificates if you are serving over HTTPS. For the database, it assumes you want to use SQLite (or you have an external database server to configure using the  installer.) It also assumes you intend to use the Filesystem driver for cache, filesystem or database driver for session, and database driver for queue (or you have an external Redis server to configure using the installer.) If you want to use other options built into Docker, see [Advanced Options](#advanced-options).

### Create compose.yml

> ```yml title="compose.yml"

Reviewer: ```suggestion
```yml {17,18} title="compose.yml"
```

services:
  panel:
    image: ghcr.io/pelican-dev/panel:latest
    restart: always
    networks:
      - default
    ports:
      - "80:80"
      - "443:443"
    extra_hosts:
      - "host.docker.internal:host-gateway"
    volumes:
      - pelican-data:/pelican-data
      - pelican-logs:/var/www/html/storage/logs
    environment:
      XDG_DATA_HOME: /pelican-data
      APP_URL: "http://localhost"
      ADMIN_EMAIL: "USEYOUROWNEMAILHERE@example.com"

volumes:
  pelican-data:
  pelican-logs:

networks:
  default:
    ipam:
      config:
        - subnet: 172.20.0.0/16
```

> ### Set Required Environment Variables
> 1. Set `APP_URL` to the base URL your server will be reachable on, including the protocol (https:// or http://) and port. 

Reviewer: ```suggestion
### Set Required Environment Variables

1. Set `APP_URL` to the base URL your panel will be reachable on, including the protocol (https:// or http://) and port. 
```

    - Note that Caddy, the integrated webserver, will serve a 308 redirect to any requests on port 80 if the `APP_URL` begins with `https://`. If your final site will be reachable over HTTPS but TLS (SSL) will be handled and terminated by an upstream server, such as a reverse proxy, you will need to use a [custom caddyfile](#custom-caddyfile).
2. Set the `ADMIN_EMAIL` to your email address. Caddy will use this email address to generate a LetsEncrypt SSL certificate if you are serving via HTTPS.

Now, close and save changes to `compose.yml`. 

> ### Starting
> From the directory in which the Caddyfile is located, run:

Reviewer: ```suggestion
### Starting

From the directory in which the compose file is located, run:
```

```sh
docker compose up -d
```

> ### Back Up Your Encryption Key
> The first time the container starts, it will generate an `APP_KEY` which is used as an encryption key. This will be saved automatically, but you should save a copy in a secure place in case you need it later.

Reviewer: ```suggestion
### Back Up Your Encryption Key

The first time the container starts, it will generate an `APP_KEY` which is used as an encryption key. This will be saved automatically, but you should save a copy in a secure place in case you need it later.
```

> Run:
> 

Reviewer: ```suggestion
```

```sh
docker compose logs panel | grep 'Generated app key:'
```

> ### Installing
> Open the installer in your browser at `APP_URL/installer` to finish setting up the panel.

Reviewer: ```suggestion
### Installing

Open the installer in your browser at `APP_URL/installer` to finish setting up the panel.
```

:::note
The first time the container starts after installing or updating, it will apply database migrations, which may take a few minutes. The panel will not be accessible during this process.
:::

> 
> #### Sensible Driver Defaults:
> 
> * Cache Driver: Filesystem
> * Database Driver: SQLite
> * Queue Driver: Database
> * Session Driver: Filesystem
> 
> For other configuration, such as UI options, CAPTCHA, email, backups and OAuth, head to the settings menu in the admin panel.

Reviewer: ```suggestion
```

Author: Can I ask why on this one? This is definitely information I would be looking for when following similar guides.

> ### Stopping
> The panel will automatically restart if the container crashes or the host restarts.  If you need to non-destructively stop the panel for any reason, navigate back to the directory containing `compose.yml` and run:

Reviewer: ```suggestion
### Stopping

The panel will automatically restart if the container crashes or the host restarts.  If you need to non-destructively stop the panel for any reason, navigate back to the directory containing `compose.yml` and run:
```

```sh
docker compose down
```

> ### Uninstalling
> To uninstall the panel, navigate to the directory containing `compose.yml` and run:

Reviewer: ```suggestion
### Uninstalling

To uninstall the panel, navigate to the directory containing `compose.yml` and run:
```

```sh
docker compose down -v
```

:::danger
    **This will permanently delete the panel and all associated data including the SQLite database and your encryption key.**
:::

> ## Advanced Options
> ### Redis
> You can deploy a Redis container in the same `compose.yml` file for easy setup. This example starts a Redis container and auto-populates the Redis configuration, cache, session and queue options in the web installer to use it.

Reviewer: ```suggestion
## Advanced Options

### Redis

You can deploy a Redis container in the same `compose.yml` file for easy setup. This example starts a Redis container and auto-populates the Redis configuration, cache, session and queue options in the web installer to use it.
```

Once you use the web installer, changes to the highlighted environment variables will not apply, and those lines can be deleted from the `compose.yml` file.

```yml title="compose.yml"
services:
  redis:
    image: redis:latest
    restart: always
    networks:
      - default

  panel:
    image: ghcr.io/pelican-dev/panel:latest
    depends_on:
      - redis
    restart: always
    networks:
      - default
    ports:
      - "80:80"
      - "443:443"
    extra_hosts:
      - "host.docker.internal:host-gateway"
    volumes:
      - pelican-data:/pelican-data
      - pelican-logs:/var/www/html/storage/logs
    environment:
      XDG_DATA_HOME: /pelican-data
      # highlight-start
      CACHE_STORE: redis
      SESSION_DRIVER: redis
      QUEUE_CONNECTION: redis
      REDIS_HOST: redis
      REDIS_PORT: 6379
      # highlight-end
      APP_URL: "http://localhost"
      ADMIN_EMAIL: "USEYOUROWNEMAILHERE@example.com"

volumes:
  pelican-data:
  pelican-logs:

networks:
  default:
    ipam:
      config:
        - subnet: 172.20.0.0/16
```

> ### Custom Caddyfile
> The default Caddyfile will work for standard installations. If you need to edit the configuration of the integrated webserver, such as to place it behind a reverse proxy that terminates TLS, you can do so by bind-mounting a Caddyfile on the host to `/etc/caddy/Caddyfile` inside the container.

Reviewer: ```suggestion
### Custom Caddyfile

The default Caddyfile will work for standard installations. If you need to edit the configuration of the integrated webserver, such as to place it behind a reverse proxy that terminates TLS, you can do so by bind-mounting a Caddyfile on the host to `/etc/caddy/Caddyfile` inside the container.
```

> # Panel in Docker

Reviewer 2: This is already in the panel directory

```suggestion
# Docker
```

Author: I agree it's redundant when you're already on the page because of the breadcrumbs (and navigating via the sidebar), but just titling the page "Docker" would make it difficult to find by searching because the built-in search bar does not show the parent directory when showing search results (plus it would help search engines to have panel and docker both in the meta title).

Pelican provides pre-built Docker images via GitHub Packages. `ghcr.io/pelican-dev/panel:latest` is the current latest release, and `ghcr.io/pelican-dev/panel:main` is built automatically from the current `main` branch. Deploying the panel in Docker is still a work in progress. While the plan is to make Docker the preferred installation method, we currently recommend the [standard deployment instructions](/docs/panel/getting-started)

> This guide requires the Docker Engine and Docker CLI. (Docker Compose has been included in the Docker CLI since v2. Docker Compose v1 is unsupported.) If you need to install them, follow the [official install instructions](https://docs.docker.com/compose/install/).

Reviewer 2: Let's keep this consistent with the other page here: https://pelican.dev/docs/wings/install/#installing-docker

Author: I moved that section to its own guide page and linked to it from both, to avoid maintaining it twice.

## Basics

> The easiest deployment method is using the standard compose.yml file.

Reviewer 2: ```suggestion
The easiest deployment method is using the standard `compose.yml` file.
```

This configuration includes an integrated web server that will automatically obtain SSL certificates if you are serving over HTTPS. For the database, it assumes you want to use SQLite (or you have an external database server to configure using the  installer.) It also assumes you intend to use the Filesystem driver for cache, filesystem or database driver for session, and database driver for queue (or you have an external Redis server to configure using the installer.) If you want to use other options built into Docker, see [Advanced Options](#advanced-options).

> ### Create compose.yml
> ```sh
> mkdir -p pelican/panel
> cd pelican/panel
> nano compose.yml
> ```
> 
> Now, paste the following into the compose.yml file:

Reviewer 2: People are too happy to copy and paste.

```suggestion
### Create compose.yml
```

> ```yml

Reviewer 2: ```suggestion
```yml title="compose.yml"
```

services:
  panel:
    image: ghcr.io/pelican-dev/panel:latest
    restart: always
    networks:
      - default
    ports:
      - "80:80"
      - "443:443"
    extra_hosts:
      - "host.docker.internal:host-gateway"
    volumes:
      - pelican-data:/pelican-data
      - pelican-logs:/var/www/html/storage/logs
    environment:
      XDG_DATA_HOME: /pelican-data
      APP_URL: "http://localhost"
      ADMIN_EMAIL: "USEYOUROWNEMAILHERE@example.com"

volumes:
  pelican-data:
  pelican-logs:

networks:
  default:
    ipam:
      config:
        - subnet: 172.20.0.0/16
```

### Set Required Environment Variables
1. Set `APP_URL` to the base URL your server will be reachable on, including the protocol (https:// or http://) and port. 

>     - Note that Caddy, the integrated webserver, will serve a 308 redirect to any requests on port 80 if the `APP_URL` begins with https://. If your final site will be reachable over HTTPS but SSL will be handled and terminated by an upstream server, such as a reverse proxy, you will need to use a [custom caddyfile](#custom-caddyfile).

Reviewer 2: ```suggestion
    - Note that Caddy, the integrated webserver, will serve a 308 redirect to any requests on port 80 if the `APP_URL` begins with `https://`. If your final site will be reachable over HTTPS but SSL will be handled and terminated by an upstream server, such as a reverse proxy, you will need to use a [custom caddyfile](#custom-caddyfile).
```

> 2. Set the `ADMIN_EMAIL` to your email address. Caddy will use this email address to generate a LetsEncrypt SSL certificate if you are serving using HTTPS.

Reviewer 2: ```suggestion
2. Set the `ADMIN_EMAIL` to your email address. Caddy will use this email address to generate a LetsEncrypt SSL certificate if you are serving via HTTPS.
```

Now, close and save changes to `compose.yml` by pressing `ctrl`+`x` then `enter`. 

> ### Start The Panel

Reviewer 2: ```suggestion
### Starting
```

From the directory in which the Caddyfile is located, run:

```sh
docker compose up -d
```

### Back Up Your Encryption Key
The first time the container starts, it will generate an `APP_KEY` which is used as an encryption key. This will be saved automatically, but you should save a copy in a secure place in case you need it later.

Run:

```sh
docker compose logs panel | grep 'Generated app key:'
```

> ### Install

Reviewer 2: ```suggestion
### Installing
```

> Open the installer in your browser at `APP_URL/installer` to finish setting up the panel.

Reviewer 2: Need to help folks out

```suggestion
Open the installer in your browser at `https://localhost/installer` to finish setting up the panel.
```

Author: I think this is the wrong approach for a couple reasons: 

- For the Docker image, you have to manually set APP_URL, so users who get to this point in the guide should understand what to put there. If we're removing commands to prevent people form copy/pasting withoug knowing what they're doing, this is no different.

- Whether or not the container itself is reachable on localhost:443, Caddy in the Docker image will not serve the panel if the URL in the request does not match the APP_URL environment variable.

> :::note
> The first time the container starts after installing or updating, it will apply database migrations. This process may take several minutes, during which the panel will not be accessible.
> :::

Reviewer 2: This takes seconds on my Mac, which is much slower than on Linux.

```suggestion
```

Author: On my server, running in Docker using MySQL with the workaround noted in https://github.com/pelican-dev/panel/issues/986, it takes a couple of minutes to apply all the migrations. I imagine it's faster on SQLite and if the schema is able to apply?

> If you don't have an external database you want to use, select SQLite with the default location for your database. If you don't have an external redis server, select Filesystem for Cache Drvier, Filesystem or Database for Session Driver, and Database for Queue Driver. 

Reviewer 2: ```suggestion
#### Sensible Driver Defaults:

* Cache Driver: Filesystem
* Database Driver: SQLite
* Queue Driver: Database
* Session Driver: Filesystem
```

> If you don't have an external database you want to use, select SQLite with the default location for your database. If you don't have an external redis server, select Filesystem for Cache Drvier, Filesystem or Database for Session Driver, and Database for Queue Driver. 

Reviewer 2: ```suggestion
If you don't have an external database you want to use, select SQLite with the default location for your database. If you don't have an external redis server, select Filesystem for Cache Driver, Filesystem or Database for Session Driver, and Database for Queue Driver. 
```

For other configuration, such as UI options, CAPTCHA, email, backups and OAuth, head to the settings menu in the admin panel.

> ### Stopping the Panel

Reviewer 2: ```suggestion
### Stopping
```

The panel will automatically restart if the container crashes or the host restarts.  If you need to non-destructively stop the panel for any reason, navigate back to the directory containing `compose.yml` and run:

```sh
docker compose down
```

### Uninstalling
To uninstall the panel, navigate to the directory containing `compose.yml` and run:

```sh
docker compose down -v
```

:::danger

>     **This will permanently delete the panel and all associated data** including the SQLite database and your encryption key.

Reviewer 2: ```suggestion
    **This will permanently delete the panel and all associated data including the SQLite database and your encryption key.**
```

:::

## Advanced Options
### Redis
You can deploy a Redis container in the same `compose.yml` file for easy setup. This example starts a Redis container and auto-populates the Redis configuration, cache, session and queue options in the web installer to use it.

Once you use the web installer, changes to the highlighted environment variables will not apply, and those lines can be deleted from the `compose.yml` file. See [Configuring with Environment Variables](#configuring-with-environment-variables).

> ```yml

Reviewer 2: ```suggestion
```yml title="compose.yml"
```

services:
  redis:
    image: redis:latest
    restart: always
    networks:
      - default

  panel:
    image: ghcr.io/pelican-dev/panel:latest
    depends_on:
      - redis
    restart: always
    networks:
      - default
    ports:
      - "80:80"
      - "443:443"
    extra_hosts:
      - "host.docker.internal:host-gateway"
    volumes:
      - pelican-data:/pelican-data
      - pelican-logs:/var/www/html/storage/logs
    environment:
      XDG_DATA_HOME: /pelican-data
      # highlight-start
      CACHE_STORE: redis
      SESSION_DRIVER: redis
      QUEUE_CONNECTION: redis
      REDIS_HOST: redis
      REDIS_PORT: 6379
      # highlight-end
      APP_URL: "http://localhost"
      ADMIN_EMAIL: "USEYOUROWNEMAILHERE@example.com"

volumes:
  pelican-data:
  pelican-logs:

networks:
  default:
    ipam:
      config:
        - subnet: 172.20.0.0/16
```

> ### MySQL
> Deploying the panel in Docker is still a work in progress, and the default Docker image does not currently support using MySQL databases. Instructions for using MySQL will be added here once it is supported.
> 

Reviewer 2: You just put in the mysql database info into the .env and it works.

```suggestion
```

Author: Not currently in the Docker image. https://github.com/pelican-dev/panel/issues/986

Author: I see that issue is fixed so I'll add docs for this soon.

### Custom Caddyfile
The default Caddyfile will work for standard installations. If you need to edit the configuration of the integrated webserver, such as to place it behind a reverse proxy that terminates TLS, you can do so by bind-mounting a Caddyfile on the host to `/etc/caddy/Caddyfile` inside the container.

This example assumes there is a Caddyfile in the same directory as the `compose.yml` file. 

> ```yml

Reviewer 2: ```suggestion
```yml {15} title="compose.yml"
```

- `APP_ENV`: Set to `production` for production deployments. `[production, local, testing]`

### Configured By Installer
By default, these environment variables are all set by the installer and saved to a .env file which is persisted in `/pelican-data/`. You can use this information to configure them when creating the container instead.

- `APP_INSTALLED`: If you are manually configuring the panel via environment variables, set it to `true` to disable the installer.
- `APP_KEY`: A 32-character-long randomly-generated string used as an encryption key.
- `APP_TIMEZONE`: The timezone for your panel. See https://www.php.net/manual/en/timezones.php for a list of options.
- `APP_NAME`: This will be the Name of your Panel. Default: Pelican
- `DB_CONNECTION`: The type of database to use for the panel. `[sqlite, mariadb, mysql]`
- `CACHE_STORE`: The driver used for caching. We recommend "Filesystem." `[file, redis]`
- `SESSION_DRIVER`: The driver used for storing sessions. We recommend "Filesystem" or "Database." `[file, database, cookie, redis]`
- `QUEUE_CONNECTION`: The driver used for handling queues. We recommend "Database." `[database, redis, sync]`

:::warning

>     If you are using `redis` for your `SESSION_DRIVER` or `QUEUE_CONNECTION`, you must set `CACHE_STORE` to `redis`.

Reviewer 3: No ?
>If you have redis might aswell use it for session, queue & cache but you can only use it for one of them if you want.

### Configured in Settings
These options are not configured by the installer but can be set in the admin settings area, or using these environment variables.

#### General
- `APP_FAVICON`: A path relative to the public directory to use as the favicon. Default: `/pelican.ico`
- `FILAMENT_TOP_NAVIGATION`: Set to `false` to use the sidebar for navigation. Set to `true` to use a top menu. `[true, false]` Default: false
- `PANEL_USE_BINARY_PREFIX`: Whether the panel should use the binary prefix to display units (MiB/GiB) or not (MB/GB). `[true, false]` Default: true
- `APP_2FA_REQUIRED`: Set to `0` to disable 2FA requirement. Set to `1` to enable for admins only. Set to `2` to enable for everyone. `[0, 1, 2]` Default: 0
- `TRUSTED_PROXIES`: A comma-separated list of trusted proxies. `*` can be used to trust any proxy that connects directly to your server, a requirement when you cannot know the address of your proxy (e.g. if using Rackspace balancers). `**` can be used to trust not just any proxy that connects directly to your server, but also proxies that connect to those proxies, and all the way back until you reach the original source IP. Default: null

> - `FILAMENT_WIDTH`: The max display width. Default: `screen-2xl`

Author: I'm not sure where to find the options for FILAMENT_WIDTH (other than in the panel, which seems to be in a different format) Is there a list somewhere I can include as options?

Reviewer 4: Its a filament enum
